### ISSUE TITLE
cpu blocked for ~10s before first request on complex Flask app with gevent on 8.10.1 release

### ISSUE THREAD

**Issue by ddorian** [newrelic/newrelic-python-agent#918]
Hi,

I have a pretty complex flask app locally that hangs for about ~10s ONLY before the first request when using 8.10.1 release (doesn't happen on 8.10.0). Still happens on 9.0.0 though. The CPU usage is 100% for those seconds. All the other requests work normally after.

How do you want me to profile it? Using ubuntu 22.04 & Python 3.11 and many libraries.

**Comment by mthipparthi** [comment]
We have the same issue. It has killed all our applications and performance.  High time it is resloved.

**Comment by TimPansino** [comment]
Can we get some more info on this issue? A pip freeze would be very helpful if you're comfortable providing it. Or at the very least gevent and greenlet versions used.

This may be related to setting `startup_timeout`. We do not recommend setting `startup_timeout` for web applications, as this blocks completion of requests to the application for the initial request(s) until registration with the backend can complete (up to the time set, in seconds). Not setting it or setting to 0 means that initial requests will not be monitored while the agent starts on a background thread. Once it completes its connect cycle, it will begin monitoring requests.

I was unable to reproduce the issue on a machine running the following app. I saw no real difference in startup time. After examining the changes between `8.10.0` and `8.10.1` the only interesting change I see is a change to prefer `importlib.metadata` calls over builtin version attributes on packages for environment reporting. This **could** have potential performance implications, but I don't see any in my initial testing. I'd like to ensure my environment is identical before I do any larger scale performance profiling sessions.

```
newrelic (8.10.0 vs 8.10.0)
Flask==2.3.3
gevent==23.9.1
greenlet==3.0.0rc3
```

```python
# Patch with gevent
from gevent import monkey
monkey.patch_all()

# Initialize agent
import newrelic.agent
newrelic.agent.initialize("newrelic.ini")

# Create application
from gevent.pywsgi import WSGIServer
from flask import Flask

app = Flask(__name__)

@app.route('/')
def hello():
    return "Hello from Flask!"

if __name__ == '__main__':
    # Start server
    print("Starting gevent server...")
    server = WSGIServer(("0.0.0.0", 8000), app)
    server.serve_forever()
```

**Comment by mthipparthi** [comment]
we run using -  `newrelic-admin run-program gunicorn --workers 3 --timeout 65 --graceful-timeout 65 --max-requests 2000 --max-requests-jitter 200 --bind 0.0.0.0:8088 app.wsgi`. same issue. Not only for gevent, even for sync workers.

**Comment by ddorian** [comment]
cc @umaannamalai

**Comment by hmstepanek** [comment]
Hey all we have a fix coming for this to be released on Monday Nov 13th.

**Comment by hmstepanek** [comment]
This fix was released in v9.1.2.

**Comment by ddorian** [comment]
@hmstepanek this is not fixed in v9.1.2 (I just tested locally).

Can you re-open the issue? Same as in the issue body (8s+ to load the first page with 100% CPU)

**Comment by hmstepanek** [comment]
@ddorian it sounds like there may be multiple issues here. We fixed the importlib metadata issue but sounds like that wasn't the root of the problem in your case. Could you try out this branch and let us know if it reduces the cpu spike: https://github.com/newrelic/newrelic-python-agent/tree/performance-cpu-increase-possible-fix. If it doesn't reduce the CPU usage, could you send us a cprofile dump of the first transaction hitting your application? Something like the following should be sufficient for profiling: `newrelic-admin run-program python -m cProfile -s tottime -o profile.dat main.py`, hit the app with the first web request, kill it, and send us the profile.dat.

**Comment by ddorian** [comment]
@hmstepanek just tested it, and it works! Thank you!

**Comment by hmstepanek** [comment]
We just released a new configuration option that you can add to you newrelic.ini config file `package_reporting.enabled = false` or disable via an env var setting `NEW_RELIC_PACKAGE_REPORTING_ENABLED=false` that allows you to disable package and version capturing information on startup of the agent (this was what was causing the CPU spike). This is available in 9.2.0. Docs will be updated later tonight with the new config option. This effectively does the same thing as that branch @ddorian

**Comment by ddorian** [comment]
Thank you
